File: chat-template-compat/popular_templates.py
import yaml
import requests
from pathlib import Path
from gguf import GGUFReader
import asyncio
import aiohttp
from typing import Optional, Dict, Any
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class ModelProcessor:
    def __init__(self):
        self.base_dir = Path(__file__).parent
        self.repos_file = self.base_dir / "tests" / "data" / "embedded-repos.txt"
        self.output_file = self.base_dir / "tests" / "data" / "embedded-repos-with-base.yaml"

    async def get_base_model(self, session: aiohttp.ClientSession, repo: str) -> Optional[str]:
        url = f"https://huggingface.co/api/models/{repo}"
        try:
            async with session.get(url) as response:
                if response.status == 200:
                    data = await response.json()
                    return data.get("cardData", {}).get("base_model")
                else:
                    logger.error(f"Failed to fetch base model for {repo}: {response.status}")
                    return None
        except Exception as e:
            logger.error(f"Error fetching base model for {repo}: {e}")
            return None

    async def get_gguf_file_path(self, session: aiohttp.ClientSession, repo: str) -> Optional[str]:
        url = f"https://huggingface.co/api/models/{repo}/tree/main"
        try:
            async with session.get(url) as response:
                if response.status == 200:
                    files = await response.json()
                    gguf_file = next((f["path"] for f in files if f["path"].endswith(".gguf")), None)
                    return gguf_file
                else:
                    logger.error(f"Failed to fetch file list for {repo}: {response.status}")
                    return None
        except Exception as e:
            logger.error(f"Error fetching GGUF file path for {repo}: {e}")
            return None

    async def get_chat_template(self, repo: str, gguf_path: str) -> Optional[str]:
        url = f"https://huggingface.co/{repo}/resolve/main/{gguf_path}"
        try:
            reader = GGUFReader(url)
            metadata = reader.metadata
            return metadata.get("tokenizer.chat_template")
        except Exception as e:
            logger.error(f"Error getting chat template for {repo}: {e}")
            return None

    async def process_repo(self, session: aiohttp.ClientSession, repo: str) -> Optional[Dict[str, Any]]:
        logger.info(f"Processing repo: {repo}")
        
        base_model = await self.get_base_model(session, repo)
        if not base_model:
            return None

        gguf_path = await self.get_gguf_file_path(session, repo)
        if not gguf_path:
            return None

        chat_template = await self.get_chat_template(repo, gguf_path)
        if not chat_template:
            return None

        return {
            "id": repo,
            "base": base_model,
            "template": chat_template
        }

    async def process_all_repos(self):
        # Read repos file
        repos = self.repos_file.read_text().strip().split('\n')
        
        async with aiohttp.ClientSession() as session:
            tasks = [self.process_repo(session, repo) for repo in repos]
            results = await asyncio.gather(*tasks)
            
            # Filter out None results and create final data
            data = [result for result in results if result]
            
            # Write YAML file
            self.output_file.parent.mkdir(parents=True, exist_ok=True)
            with self.output_file.open('w') as f:
                yaml.dump(data, f, sort_keys=False, allow_unicode=True)
            
            logger.info(f"Processed {len(data)} repos successfully")

def main():
    processor = ModelProcessor()
    asyncio.run(processor.process_all_repos())

if __name__ == "__main__":
    main() 

File: chat-template-compat/popular-templates.js
import { gguf } from "@huggingface/gguf";
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import yaml from 'js-yaml';

// Get current script directory
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Compute output directory relative to script location
const OUTPUT_DIR = path.join(__dirname, 'tests', 'data', 'embedded');
const REPOS_FILE = path.join(__dirname, 'tests', 'data', 'embedded-repos.txt');
const OUTPUT_YAML = path.join(__dirname, 'tests', 'data', 'embedded-repos-with-base.yaml');

async function ensureDirectoryExists(dirPath) {
  try {
    await fs.mkdir(dirPath, { recursive: true });
  } catch (error) {
    if (error.code !== 'EEXIST') throw error;
  }
}

async function fileExists(filePath) {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function fetchOwnerModels(owner, limit = 500) {
  let allModels = [];
  const url = `https://huggingface.co/api/models?author=${owner}&limit=${limit}`;
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to fetch models for ${owner}: ${response.statusText}`);
  }
  const models = await response.json();
  const filteredModels = models
    .filter(model => model.modelId.startsWith(owner));
  allModels.push(...filteredModels.map(model => model.modelId));
  return allModels;
}

async function fetchFirstGGUFFile(owner_repo) {
  const url = `https://huggingface.co/api/models/${owner_repo}/tree/main`;
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to fetch file list for ${owner_repo}: ${response.statusText}`);
  }

  const files = await response.json();
  const ggufFile = files.find(file => file.path.endsWith('.gguf'));

  if (!ggufFile) {
    throw new Error(`No GGUF file found in ${owner_repo}`);
  }

  return ggufFile.path;
}

async function fetchBaseModel(repo) {
  const url = `https://huggingface.co/api/models/${repo}`;
  try {
    const response = await fetch(url);
    if (!response.ok) {
      console.error(`Failed to fetch base model for ${repo}: ${response.statusText}`);
      return null;
    }
    const data = await response.json();
    return data?.cardData?.base_model;
  } catch (error) {
    console.error(`Error fetching base model for ${repo}:`, error);
    return null;
  }
}

async function getModelInfo(repo) {
  try {
    console.log(`Processing ${repo}...`);
    
    const baseModel = await fetchBaseModel(repo);
    if (!baseModel) {
      console.log(`No base model found for ${repo}`);
      return null;
    }

    const filename = await fetchFirstGGUFFile(repo);
    const url = `https://huggingface.co/${repo}/resolve/main/${filename}`;
    console.log(`Reading GGUF file from ${url}...`);

    const { metadata } = await gguf(url);
    const template = metadata?.['tokenizer.chat_template'];
    
    if (!template) {
      console.log(`No chat template found for ${repo}/${filename}`);
      return null;
    }

    return {
      id: repo,
      base: baseModel,
      template: template
    };
  } catch (error) {
    console.error(`Error processing ${repo}:`, error);
    return null;
  }
}

async function main() {
  const owner = 'lmstudio-community';
  try {
    const modelRepos = await fetchOwnerModels(owner);
    console.log(`Found ${modelRepos.length} model repos for ${owner}`);

    // Save repos list to file
    await ensureDirectoryExists(path.dirname(REPOS_FILE));
    await fs.writeFile(REPOS_FILE, modelRepos.join('\n'));
    console.log(`Saved repos list to ${REPOS_FILE}`);

    // Process all repos in parallel with concurrency limit
    const concurrencyLimit = 5;
    const results = [];
    
    for (let i = 0; i < modelRepos.length; i += concurrencyLimit) {
      const batch = modelRepos.slice(i, i + concurrencyLimit);
      const batchResults = await Promise.all(batch.map(repo => getModelInfo(repo)));
      results.push(...batchResults.filter(Boolean));
    }

    // Save results to YAML file
    await ensureDirectoryExists(path.dirname(OUTPUT_YAML));
    await fs.writeFile(
      OUTPUT_YAML,
      yaml.dump(results, { lineWidth: -1, noRefs: true })
    );
    
    console.log(`Successfully processed ${results.length} models`);
    console.log(`Results saved to ${OUTPUT_YAML}`);
  } catch (error) {
    console.error('Error in main:', error);
  }
}

main(); 